nrpc.go

```diff
@@ -512,3 +704,185 @@ func (k *KeepStreamAlive) loop() {
+			if deadline.After(now) {
+				// Safety call to setupStreamedReply in case QueueRequest had
+				// to time to do it yet
+				request.setupStreamedReply()
+				log.Printf("scheduler: try scheduling")
+				select {
+				case pool.schedule <- request:
+					log.Printf("scheduler: scheduled the request")
+					continue queueLoop
+				case <-time.After(deadline.Sub(now)):
+					// Too late
+					log.Printf("scheduler: could not schedule the request")
+				}
+			} else {
+				log.Printf("scheduler: already too late, skip scheduling")
+			}
+
+			log.Printf("Sending SERVERTOOBUSY to the caller")
+			request.SendErrorTooBusy("No worker available")
+		}
+	}
+}
+
+func (pool *WorkerPool) worker() {
+	defer pool.waitGroup.Done()
+	for request := range pool.schedule {
+		if request == nil {
+			log.Printf("worker: got nil, exiting")
+			return
+		}
+		log.Printf("worker: got a request, running it")
+		request.RunAndReply()
+	}
+}
+
+// SetMaxPending changes the queue size
+func (pool *WorkerPool) SetMaxPending(value uint) {
+	if pool.maxPending == value {
+		return
+	}
+	pool.m.Lock()
+	defer pool.m.Unlock()
+
+	oldQueue := pool.queue
+	pool.queue = make(chan *Request, value)
+	pool.maxPending = value
+
+	close(oldQueue)
+
+	// drain the old queue and cancel requests if there are too many
+	for request := range oldQueue {
+		select {
+		case pool.queue <- request:
+		default:
+			request.SendErrorTooBusy("too many pending requests")
+		}
+	}
+}
+
+// SetMaxPendingDuration changes the max pending delay
+func (pool *WorkerPool) SetMaxPendingDuration(value time.Duration) {
+	pool.m.Lock()
+	pool.maxPendingDuration = value
+	pool.m.Unlock()
+}
+
+// SetSize changes the number of workers
+func (pool *WorkerPool) SetSize(size uint) {
+	pool.m.Lock()
+	defer pool.m.Unlock()
+
+	if size == pool.size {
+		return
+	}
+	for size < pool.size {
+		pool.schedule <- nil
+		pool.size--
+	}
+	for size > pool.size {
+		pool.waitGroup.Add(1)
+		go pool.worker()
+		pool.size++
+	}
+}
+
+// QueueRequest adds a request to the queue
+// Send a SERVERTOOBUSY error to the client if the queue is full
+func (pool *WorkerPool) QueueRequest(request *Request) error {
+	select {
+	case pool.getQueue() <- request:
+		request.setupStreamedReply()
+		return nil
+	default:
+		return request.SendErrorTooBusy("too many pending requests")
+	}
+}
+
+// Close stops all the workers and wait for their completion
+// If the workers do not stop before the timeout, their context is canceled
+// Will never return if a request ignore the context
+func (pool *WorkerPool) Close(timeout time.Duration) {
+	// Stops all the workers so nothing more gets scheduled
+	pool.SetSize(0)
+
+	pool.m.Lock()
+	oldQueue := pool.queue
+	pool.queue = nil
+	pool.m.Unlock()
+
+	close(oldQueue)
+	for request := range oldQueue {
+		request.SendErrorTooBusy("Worker pool shutting down")
+	}
+
+	// Now wait for the workers to stop and cancel the context if they don't
+	timer := time.AfterFunc(timeout, pool.contextCancel)
```

Comment:
See #51

Reply:
So when we get to line 884, it is ensured that no more jobs will be picked up by workers. There are still some workers running that are executing handlers, and the handlers have been given `ctx`. Ideally, the handlers are in turn passing down `ctx` to it's children. Once each handler exits, it's worker will also exit.

If we were to now cancel `ctx`, by calling `pool.contextCancel()`, the whole of the context tree should get cancelled. In theory. So, after initiating the cancellation by calling `pool.contextCancel()`, we wait for a while -- up to `timeout` -- to let it take effect.

If the timeout was 5 seconds, and the workers actually cleaned up in 1 second, then `pool.waitGroup.Wait()` will complete in 1 second.

This is my understanding. Do you agree?

Reply:
> So when we get to line 884, it is ensured that no more jobs will be picked up by workers. There are still some workers running that are executing handlers, and the handlers have been given `ctx`. Ideally, the handlers are in turn passing down `ctx` to it's children. Once each handler exits, it's worker will also exit.
> 
Exactly

> If we were to now cancel `ctx`, by calling `pool.contextCancel()`, the whole of the context tree should get cancelled. In theory. So, after initiating the cancellation by calling `pool.contextCancel()`, we wait for a while -- up to `timeout` -- to let it take effect.

If you mean contextCancel is called after waiting for the effective timeout, then yes. The thing is that the timeout is not a timeout for the complete Close() function. It's a timeout for handlers to finish normally. If some workers don't stop quickly enough, 

> 
> If the timeout was 5 seconds, and the workers actually cleaned up in 1 second, then `pool.waitGroup.Wait()` will complete in 1 second.
> 

Yes, and cancelling the context becomes irrelevant.

Reply:
> If you mean contextCancel is called after waiting for the effective timeout

Well, no. I mean that we haven't started the process of cancelling the context tree. Only after we do that, will the user's handlers get a chance to abort. For example, if a worker is running this handler:

```go
func (s *server) SayHello(ctx context.Context, req helloworld.HelloRequest) (resp helloworld.HelloReply, err error) {
    // ..
    conn.QueryContext(ctx, "SELECT * FROM blah")
    // ..
    return
}
```

Now if I **first** cancel the context by calling `pool.contextCancel()`, then:
1. the `conn.QueryContext` call will abort with a "context cancelled" error
2. the handler will return
3. the worker will return

And we wait for up to `timeout` duration for steps 1-3 to complete.

Right now the code waits unconditionally for timeout, then cancels. I'm saying we need to cancel, then wait conditionally for up to timeout.

Reply:
This is the behavior I purposefully implemented.

I understand your point of view, which is to unconditionally cancel the context so the workers stops.
With this approach the timeout becomes the maximum time we wait for the workers to stop, after what we should just leave and let the workers die alone.

This is not the way I chose
First, I want the workers that are almost done to finish by themselves.
If it's too long, I use the context to force the workers to stop without reporting to the caller. It possible I want to avoid using the context to stop the handlers. Said differently, canceling the context is a last resort option.

The reason I chose to do that (and not directly canceling the context) is that every running worker means there is a client waiting for an answer. And I prefer my server to shutdown gracefully, ie respond to the clients successfully if possible in a reasonable delay.

If someone don't want to wait and cancel directly, it's possible by calling Close(0).

I also think that a stalled worker should stall the pool and block the Close(), ie the Wait() must complete.

If I cannot convince you then I need to find another way to implement a graceful shutdown of the pool.

Let me know what you think.

Reply:
Will the current code work correctly if there is a handler that runs for say `2*time.Minute` and `pool.Close(time.Minute)` is called? When is it safe for the user to close the nats connection?

I'm OK with the current approach, since it is a conscious and reasonable design decision, and the code is aligned with the design.

Reply:
If the handler runs for 2 minutes, and ignore the context cancel, then pool.Close(time.Minute) will return after 2 minutes. Then we can can safely close the nats connection.

No matter what, if pool.Close() returns, it is safe to close the connection.

If it doesn't, the developer needs to kill -QUIT the process and find the handler routine that ignored the context cancel.

Thanks for approving, I will merge the request.
Some smaller changes will follow, in their own PRs if they need review.